### PR TITLE
feat(query): initial implementation of view query.

### LINKED ISSUE #1935: Support ViewQuery in addition to Query, clean up work
# Goals
- [x] `@Query(Type, {descendants: true})` currently queries light DOM
- [ ] `@ViewQuery(Type, {descendants: true})` should query view DOM
- [x] `QueryList<Type>` should be parametrizable.
- [x] `QueryList.length` length should be defined on `QueryList` as it is convenient to bind to.
- [x] `QueryList.first|last` should have `first` and `last` convenience methods on `QueryList`

### PR DESCRIPTION
**Pull request by rkirov**
ViewQuery is a new API that allows a component to query its view.

Closes #1935

### CODE REVIEW

File: modules/angular2/test/core/compiler/query_integration_spec.ts [hunk 1]
```diff
@@ -273,6 +273,86 @@ export function main() {
...
+           var template = '<needs-view-query #q><div text="ignoreme"></needs-view-query>';
+
+           tcb.overrideTemplate(MyComp, template)
+               .createAsync(MyComp)
+               .then((view) => {
+                 var q: NeedsViewQuery = view.componentViewChildren[0].getLocal("q");
+
+                 view.detectChanges();
+
+                 expect(q.query.length).toBe(3);
```
Comment (vsavkin) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: You are checking that the length is 3 and then verifying the first and the last elements are correct. What about the second one? I think it's better to do something like:

```
expect(q.query.map(s => s.text)).toEqual(["1", "2", "3"])
```

File: modules/angular2/test/core/compiler/query_integration_spec.ts [hunk 2]
```diff
@@ -321,6 +401,45 @@ class NeedsQueryByTwoLabels {
   }
 }
 
+@Component({selector: 'needs-view-query'})
+@View({
+  directives: [NgFor, TextDirective],
+  template: '<div text="1"><div text="need descendants"></div></div>' +
+                '<div *ng-for="var i of [\'2\', \'3\']" [text]="i"></div>'
```
Comment (vsavkin) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: Do you use ng-for to check some ng-for specific behavior? Cause `<div [text]="2"><div [text]="3">` reads nicer.

File: modules/angular2/test/core/compiler/query_integration_spec.ts [hunk 3]
```diff
@@ -273,6 +273,86 @@ export function main() {
...
+
+                 expect(q.query.length).toBe(4);
+                 expect(q.query.first.text).toEqual("1");
+                 expect(q.query.last.text).toEqual("4");
+
+                 async.done();
+               });
+         }));
+
+      it('should include directive present on the host element',
```
Comment (vsavkin) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: This behavior is surprising to me. Imo the host element is not part of the view. And you can already query the host element using regular queries.

Realistically, I think there are three scenarios:
- Querying "self"
- Querying light dom children
- Querying view children

I cannot come up with a scenario where you would want to do "self + light dom" or "self + shadow dom".

File: modules/angular2/test/core/compiler/query_integration_spec.ts (hunk 1 above)
Comment (rkirov) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: Sounds good. I copied the pattern from your locals tests, but indeed thats' better.

File: modules/angular2/test/core/compiler/query_integration_spec.ts (hunk 2 above)
Comment (rkirov) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: I will extract it into its own test, dynamically created and destroyed directives need to be checked.

File: modules/angular2/test/core/compiler/query_integration_spec.ts (hunk 3 above)
Comment (rkirov) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: The semantics of `descendants` get goofy when you don't consider the host as distance 0.

I will merge as it is and we can discuss the visibility options when I work on https://github.com/angular/angular/issues/2603 on Monday.

Comment (vsavkin) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: Sounds good.

**Pull request closed without merging**